### PR TITLE
bulk import support for parquet files

### PR DESCRIPTION
**Pull request by tgruben**

### CODE REVIEW

File: sql3/planner/compilebulkinsert.go [hunk 1]
```diff
@@ -203,6 +203,14 @@ func (p *ExecutionPlanner) analyzeBulkInsertStatement(stmt *parser.BulkInsertSta
 				return sql3.NewErrIntegerLiteral(im.MapExpr.Pos().Line, im.MapExpr.Pos().Column)
 			}
 		}
+	case "PARQUET":
+		// for paquet the map expressions need to be string values
```
Comment (travisturner) [inline sql3/planner/compilebulkinsert.go:207]:
```suggestion
		// for parquet the map expressions need to be string values
```

File: sql3/planner/compilebulkinsert.go [hunk 2]
```diff
@@ -203,6 +203,14 @@ func (p *ExecutionPlanner) analyzeBulkInsertStatement(stmt *parser.BulkInsertSta
...
+		// that represent the offsets in the source file
+		for _, im := range stmt.MapList {
+			if !(im.MapExpr.IsLiteral() && typeIsString(im.MapExpr.DataType())) {
+				return sql3.NewErrIntegerLiteral(im.MapExpr.Pos().Line, im.MapExpr.Pos().Column)
```
Comment (travisturner) [inline sql3/planner/compilebulkinsert.go:211]: I don't quite understand the "must be string" logic, but does that mean this error should be `NewErrStringLiteral`?

File: sql3/sql_complex_test.go [hunk 3]
```diff
@@ -4,13 +4,21 @@ package sql3_test
 import (
 	"context"
 	"fmt"
+	"io/ioutil"
```
Comment (travisturner) [inline sql3/sql_complex_test.go:7]: you probably got this warning, but this package is deprecated.

File: sql3/sql_complex_test.go [hunk 4]
```diff
@@ -2631,3 +2638,216 @@ func TestPlanner_BulkInsert_FB1831(t *testing.T) {
...
+	return nil
+}
+
+func TestPlanner_BulkInsertParquet_File(t *testing.T) {
+	// check that can pull parquet file from URL and load
+	c := test.MustRunCluster(t, 1)
+	defer c.Close()
+
+	_, _, err := sql_test.MustQueryRows(t, c.GetNode(0).Server, `create table j1 (_id ID, a INT, b DECIMAL(2), c STRING);`)
+	if err != nil {
```
Comment (travisturner) [inline sql3/sql_complex_test.go:2695]: I would prefer we use testify `assert.NoError(t, err)` (along with any other testify methods) in tests going forward.

File: sql3/sql_complex_test.go [hunk 5]
```diff
@@ -2631,3 +2638,216 @@ func TestPlanner_BulkInsert_FB1831(t *testing.T) {
...
 	}
 }
+
+type tb struct {
+	Name  string
+	Type  arrow.DataType
+	Value interface{}
+}
+
+func simpleParquetMaker(f *os.File, numRows int64, input []tb) error {
```
Comment (travisturner) [inline sql3/sql_complex_test.go:2648]: we could make this a `t.Helper()`, use `assert.NoError(t, err)` within, and don't bother returning an error on this helper function

**Review by travisturner — COMMENTED** [review]
This looks pretty slick.
I know you didn't ask for my review, but I wanted to look through it, and ended up leaving some comments.

One other thing I wondered is if we could share the same `MustRunCluster` for all three of these tests to avoid adding more startups to the tests. Is there a risk with that?

File: sql3/planner/compilebulkinsert.go (hunk 2 above)
Comment (tgruben) [reply sql3/planner/compilebulkinsert.go:211]: good ole copy paste get's me every time

File: sql3/sql_complex_test.go (hunk 4 above)
Comment (tgruben) [reply sql3/sql_complex_test.go:2695]: forgot all about testify, will update

**Comment by travisturner** [comment]
I think I might have misunderstood how `MustRunCluster()` works. Is it already shared? You may have had it configured correctly to begin with. Sorry.

**Comment by tgruben** [comment]
thanks @travisturner , your suggestions made it look a lot cleaner

**Review by travisturner — APPROVED** [review]
I just imported my first parquet file!

The only thing I would add, and it's very minor, is that you could use the `TempFile` function in the `featurebase/testhook` package to create the temp files, and then you don't have to explicitly remove them.

But approved either way.

**Pull request merged**